modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/CalciteQueryProcessor.java

```diff
@@ -530,23 +531,28 @@ private <T> List<T> parseAndProcessQuery(
 
         List<T> res = new ArrayList<>(qryList.size());
         List<RootQuery<Object[]>> qrys = new ArrayList<>(qryList.size());
+        int qryIdx = 0;
 
         for (final SqlNode sqlNode: qryList) {
+            int qryIdx0 = qryIdx;
```

Comment:
Looks like it's enough to validate count of parameters just after parsing: Something like:
```
        int[] maxDynParam = new int[] {-1};

        qryList.accept(new SqlShuttle() {
            @Override public SqlNode visit(SqlDynamicParam param) {
                if (param.getIndex() > maxDynParam[0])
                    maxDynParam[0] = param.getIndex();

                return param;
            }
        });

        int paramsCnt = params == null ? 0 : params.length;

        if (paramsCnt != maxDynParam[0] + 1)
            throw new IgniteSQLException("Wrong number of query parameters...", IgniteQueryErrorCode.PARSING);
```
And no need to change Validator and pass additional parameters to Query, PlanningContext, etc.